experimental/traffic-portal/src/app/core/users/roles/detail/role-detail.component.ts

```diff
@@ -0,0 +1,110 @@
+/*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import { Location } from "@angular/common";
+import { Component, OnInit } from "@angular/core";
+import { MatDialog } from "@angular/material/dialog";
+import { ActivatedRoute } from "@angular/router";
+import { ResponseRole } from "trafficops-types";
+
+import { UserService } from "src/app/api";
+import { DecisionDialogComponent } from "src/app/shared/dialogs/decision-dialog/decision-dialog.component";
+import { NavigationService } from "src/app/shared/navigation/navigation.service";
+
+/**
+ * AsnDetailComponent is the controller for the ASN add/edit form.
+ */
+@Component({
+	selector: "tp-role-detail",
+	styleUrls: ["./role-detail.component.scss"],
+	templateUrl: "./role-detail.component.html"
+})
+export class RoleDetailComponent implements OnInit {
+	public new = false;
+	public permissions = "";
+	public role!: ResponseRole;
+	constructor(private readonly route: ActivatedRoute, private readonly userService: UserService,
+		private readonly location: Location, private readonly dialog: MatDialog,
+		private readonly header: NavigationService) {
+	}
+
+	/**
+	 * Angular lifecycle hook where data is initialized.
+	 */
+	public async ngOnInit(): Promise<void> {
+		const role = this.route.snapshot.paramMap.get("name");
+		if (role === null) {
+			console.error("missing required route parameter 'name'");
+			return;
+		}
+
+		if (role === "new") {
```

Comment:
This means that I can never create a Role named `new`. Even worse, if such a Role were created either through TPv1 or the API, I'd be unable to edit it.

You should instead add a special route that takes _no_ parameters (e.g. `new-role`) and when the `name` route parameter is `null` you know to create a new Role. It's actually somewhat easier than the handling that needs to be done for ID-based objects, since it's got one fewer unreachable error states (technically; the component can be incorrectly embedded in a view where it may load on the wrong route, but we don't do that here). 

For an example of this, see the Capability component I added.